### ISSUE TITLE
Add gl::ScopedDepthBuffer( enableRead, enableWrite )?

### ISSUE THREAD

**Issue by paulhoux** [cinder/Cinder#824]
This is more a reminder that I want to add this to our RAII objects.

**Comment by richardeakin** [comment]
I'd like this too, and as convenience an overload that only takes one bool that enables or disables both read and write.

**Comment by simongeilfus** [comment]
Haha well, that's a funny coincidence... I just pushed [that branch](https://github.com/simongeilfus/Cinder/tree/glNext_ScopedDepthTesting)!
Didn't had time to fully test it yet though

**Comment by simongeilfus** [comment]
I just pushed a commit that changed the class to the name Paul was suggesting, I guess ScopedDepthTesting was a bit confusing because of GL_DEPTH_TEST.

What about adding also a gl::ScopedDepthFunc? All the stack/push/pop needed are already in there, and unfortunately as if we have a gl::ScopedDepthBuffer( bool ) it's impossible to have a gl::ScopedDepthBuffer( GLenum func ) without introducing ambiguity. So we'll need another class if we want to have this other convenience. I personally change a lot the depth function without touching at the other states.

**Comment by simongeilfus** [comment]
the current interface is like this:
```c++
gl::ScopedDepthBuffer enableAll;
gl::ScopedDepthBuffer disableDepth( false );
gl::ScopedDepthBuffer disableDepthWrite( true, false );
gl::ScopedDepthBuffer enableReadWriteWithGreater( true, true, GL_GREATER );
```

And could look like this for the depth func:

```c++
gl::ScopedDepthFunc depthFunc( GL_LEQUAL );
```

**Comment by simongeilfus** [comment]
Would that be better?

~~gl::ScopedDepthBuffer enableAll;~~
```c++
gl::ScopedDepthBuffer disableBothReadAndWrite( false );
gl::ScopedDepthBuffer enableBothReadAndWritel( true );
gl::ScopedDepthBuffer disableDepthWrite( true, false );
gl::ScopedDepthBuffer enableReadWriteWithGreater( true, true, GL_GREATER );
```

**Comment by richardeakin** [comment]
I like your current proposed version, handling all three cases with constructor overloads. Yea not sure if we need a default enable = true, the following seems like it would read quite nicely to me:

```cpp
gl::ScopedDepthBuffer scopedDepth( true );
```

Not that I'm against a default bool = true, either.

**Comment by simongeilfus** [comment]
Ok cool, I just added some more comments and removed the default enable = true, as the only gl::Scoped* constructors that doesn't take any argument seem to only push the state to the stack without modifying anything. Which is not the case here...

Just sent a PR for this, let me know if you'd like to change anything!

Cheers!

**Comment by richardeakin** [comment]
Resolved with #845

**Comment by paulhoux** [comment]
I like how this turned out. Good work, guys. Thanks!
